### PR TITLE
Redirect support in Swift driver

### LINKED ISSUE #800: Redirect support in Swift driver
Currently registry is serving content directly in case SWIFT is used as storage.

It will be nice if Swift driver can utilize tempUrl functionaly to offload the download to Swift directly.

### PR DESCRIPTION
**Pull request by lebauce**
Fixes #800

### CODE REVIEW

**Comment by gschukin** [comment]
@lebauce can you please document AccessKey and SecretKey params

**Comment by lebauce** [comment]
@gschukin Sure, done.

**Comment by gschukin** [comment]
@lebauce Thanks for update.

Can you please also update tempurlmethods description. Currently it is unclear what options are allowed. Have to look at lebauce#2 for clarification

File: registry/storage/driver/swift/swift.go [hunk 1]
```diff
@@ -176,11 +189,65 @@ func New(params Parameters) (*Driver, error) {
...
+		if err := mapstructure.Decode(config, &info); err == nil {
+			d.TempURLContainerKey = info.Swift.Version >= "2.2.2"
+			d.TempURLMethods = info.Tempurl.Methods
+		}
+	} else {
+		d.TempURLContainerKey = params.TempURLContainerKey
+		d.TempURLMethods = params.TempURLMethods
+	}
+
+	if len(d.TempURLMethods) > 0 {
```
Comment (gschukin) [inline registry/storage/driver/swift/swift.go]: According to HP Helion documentation there is no need to define X-Account-Meta-Temp-URL-Key
```
There are two differences between Object Storage and OpenStack Swift TempURL signature generation:

OpenStack Swift Temporary URLs (TempURL) required the X-Account-Meta-Temp-URL-Key header be set on the Swift account. In Object Storage you do not need to do this. Instead we use Access Keys to provide similar functionality.

Object Storage Temporary URLs require the user's Project ID and Access Key ID to be prepended to the signature. OpenStack Swift does not.
```

So I assume additional check should be addded to check AccessKey == ""

**Comment by gschukin** [comment]
According to Swift changelog, container TempUrl keys feature was added in 2.3.0 not in 2.2.2
https://github.com/openstack/swift/blob/master/CHANGELOG#L184

File: registry/storage/driver/swift/swift.go [hunk 2]
```diff
@@ -590,9 +657,46 @@ func (d *driver) Delete(ctx context.Context, path string) error {
...
+	methodString := "GET"
+	method, ok := options["method"]
+	if ok {
+		if methodString, ok = method.(string); !ok {
+			return "", storagedriver.ErrUnsupportedMethod
+		}
+	}
+
+	if d.AccessKey != "" && methodString == "HEAD" {
+		methodString = "GET"
```
Comment (BrianBland) [inline registry/storage/driver/swift/swift.go]: Is there a reason why we sign HEAD requests with GET?

Comment (lebauce) [inline registry/storage/driver/swift/swift.go]: Yes, I should probably add a comment about it. On HP Cloud (the `AccessKey` parameter is only used on HP Cloud), a "HEAD" on a temporary URL is allowed only if the signature was generated with "GET" or "POST".

**Comment by lebauce** [comment]
@BrianBland It seems the CircleCI tests are failing on the URLFor test. Is it running the test suite against a real Swift server ?

**Comment by RichardScothern** [comment]
@lebauce : please rebase to get the fixes for the golint build breakage.

The URLFor function in the test is returning `This server could not verify that you are authorized to access the document you requested`.  

It looks like the unit tests run against a mock swift implementation (defined in `swift_test.go`) - I see not swift environement variables on our CI server.

**Comment by RichardScothern** [comment]
ping @lebauce

**Comment by lebauce** [comment]
@gschukin Fixed both the wrong Swift version check and the doc

**Comment by RichardScothern** [comment]
Thanks @lebauce .  Can you squash the commits please?

@gschukin : does this look ok with you?

**Comment by lebauce** [comment]
@RichardScothern Done

File: registry/storage/driver/swift/swift.go [hunk 3]
```diff
@@ -590,9 +657,48 @@ func (d *driver) Delete(ctx context.Context, path string) error {
...
+			supported = true
+			break
+		}
+	}
+
+	if !supported {
+		return "", storagedriver.ErrUnsupportedMethod
+	}
+
+	expiresTime := time.Now().Add(20 * time.Minute).Unix()
```
Comment (BrianBland) [inline registry/storage/driver/swift/swift.go]: Could you support the `expiry` option like the s3 driver?

File: registry/storage/driver/swift/swift.go [hunk 4]
```diff
@@ -590,9 +657,48 @@ func (d *driver) Delete(ctx context.Context, path string) error {
...
+	methodString := "GET"
+	method, ok := options["method"]
+	if ok {
+		if methodString, ok = method.(string); !ok {
+			return "", storagedriver.ErrUnsupportedMethod
+		}
+	}
+
+	if d.AccessKey != "" && methodString == "HEAD" {
+		// On HP Cloud, a "HEAD" on a temporary URL is allowed only if the
```
Comment (BrianBland) [inline registry/storage/driver/swift/swift.go]: :+1: From [the official OpenStack docs](http://docs.openstack.org/liberty/config-reference/content/object-storage-tempurl.html) this should be the case for any reference Swift implementation.

File: registry/storage/driver/swift/swift.go (hunk 3 above)
Comment (lebauce) [inline registry/storage/driver/swift/swift.go]: Done

File: registry/storage/driver/swift/swift.go (hunk 4 above)
Comment (lebauce) [inline registry/storage/driver/swift/swift.go]: Indeed. I fixed it but the fake Swift server the testsuite is run against needed to be patched. A PR is on the way (https://github.com/ncw/swift/pull/56).

**Comment by RichardScothern** [comment]
hi @lebauce this LGTM.  Is this ready to merge?

**Comment by gschukin** [comment]
@RichardScothern LGTM
Will try this under heavy load in a couple of days

**Pull request merged**